emmarin/cl/cl/src/cl/mmr.rs

```diff
@@ -0,0 +1,126 @@
+use crate::cl::merkle;
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct MMR {
+    pub roots: Vec<Root>,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct Root {
+    pub root: [u8; 32],
+    pub height: u8,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub struct MMRProof {
+    pub path: Vec<merkle::PathNode>,
+}
+
+impl MMR {
+    pub fn new() -> Self {
+        Self { roots: vec![] }
+    }
+
+    pub fn push(&mut self, elem: &[u8]) -> MMRProof {
+        let new_root = Root {
+            root: merkle::leaf(elem),
+            height: 1,
+        };
+        self.roots.push(new_root);
+
+        let mut path = vec![];
+
+        for i in (1..self.roots.len()).rev() {
+            if self.roots[i].height == self.roots[i - 1].height {
+                path.push(merkle::PathNode::Left(self.roots[i - 1].root));
+
+                self.roots[i - 1] = Root {
+                    root: merkle::node(self.roots[i - 1].root, self.roots[i].root),
+                    height: self.roots[i - 1].height + 1,
+                };
+
+                self.roots.remove(i);
+            } else {
+                break;
+            }
+        }
+
+        MMRProof { path }
+    }
+
+    pub fn verify_proof(&self, elem: &[u8], proof: &MMRProof) -> bool {
+        let path_len = proof.path.len();
+        let leaf = merkle::leaf(elem);
+        let root = merkle::path_root(leaf, &proof.path);
+
+        for mmr_root in self.roots.iter() {
+            if mmr_root.height == (path_len + 1) as u8 {
+                return mmr_root.root == root;
+            }
+        }
+
+        false
+    }
+
+    pub fn commit(&self) -> [u8; 32] {
+        let mut hasher = Sha256::new();
+        for mrr_root in self.roots.iter() {
```

Comment:
A few spots in this file where we have the typo "mrr" instead of "mmr"

Reply:
i must have messed up the first and the autocomplete did the rest